### PR TITLE
Use MD5 for creating cache keys

### PR DESCRIPTION
**Pull request by bsudekum**
_Beginning to fix https://github.com/mapbox/mapbox-navigation-ios/issues/1391_

As pointed out by @akitchen, `cacheKeyForKey` is most likely too naive. 

https://github.com/mapbox/mapbox-navigation-ios/blob/d7f8c4d075ca578983cbe02d819d74eb7ac284c4/MapboxNavigation/Cache.swift#L122-L128

- [x]  add failing test case
- [x] actually fix code

/cc @mapbox/navigation-ios

### CODE REVIEW

File: MapboxNavigation/Cache.swift [hunk 1]
```diff
@@ -109,22 +109,22 @@ internal class FileCache {
...
 
-    private func cacheKeyForKey(_ key: String) -> String {
+    func cacheKeyForKey(_ key: String) -> String {
         if let keyAsURL = URL(string: key) {
-            return String.init(keyAsURL.lastPathComponent.hashValue)
+            return MD5(keyAsURL.lastPathComponent)
         }
 
-        return String.init(key.hashValue)
+        return MD5(key)
```
Comment (bsudekum) [inline MapboxNavigation/Cache.swift]: This single line is the actual fix. As pointed out by @akitchen, there are some shortcuts String.hashValue takes some shortcuts and things in the middle of the string may not impact the hash.

File: MapboxNavigation/Cache.swift [hunk 2]
```diff
@@ -109,22 +109,22 @@ internal class FileCache {
...
+    func cacheURLWithKey(_ key: String) -> URL {
         let cacheKey = cacheKeyForKey(key)
         return diskCacheURL.appendingPathComponent(cacheKey)
     }
 
-    private func cacheKeyForKey(_ key: String) -> String {
+    func cacheKeyForKey(_ key: String) -> String {
         if let keyAsURL = URL(string: key) {
-            return String.init(keyAsURL.lastPathComponent.hashValue)
+            return MD5(keyAsURL.lastPathComponent)
```
Comment (akitchen) [inline MapboxNavigation/Cache.swift]: I guess I'm surprised to see us still keying off of `lastPathComponent` -- I thought this was part of what was biting us. (?)

File: MapboxNavigation/Cache.swift (hunk 1 above)
Comment (akitchen) [inline MapboxNavigation/Cache.swift]: Do voice instructions fall through this code path? Asking for a friend...

File: MapboxNavigation/Cache.swift (hunk 2 above)
Comment (bsudekum) [reply MapboxNavigation/Cache.swift]: Not that I am aware of. It really only came down to two different strings having the same hashValue: https://github.com/mapbox/mapbox-navigation-ios/issues/1391#issuecomment-387490015

Comment (bsudekum) [reply MapboxNavigation/Cache.swift]: I'm happy to remove this check though. I think creating an md5 of the `key` here would be okay.

File: MapboxNavigationTests/DataCacheTests.swift [hunk 3]
```diff
@@ -106,4 +106,27 @@ class DataCacheTests: XCTestCase {
 
         NotificationCenter.default.post(name: .UIApplicationDidReceiveMemoryWarning, object: nil)
     }
+    
+    func testCacheKeyForKey() {
+        let threeMileInstruction = "<speak><amazon:effect name=\"drc\"><prosody rate=\"1.08\">Continue on <say-as interpret-as=\"address\">I-80</say-as> East for 3 miles</prosody></amazon:effect></speak>"
+        let sixMileInstruction = "<speak><amazon:effect name=\"drc\"><prosody rate=\"1.08\">Continue on <say-as interpret-as=\"address\">I-80</say-as> East for 6 miles</prosody></amazon:effect></speak>"
+        XCTAssertNotEqual(cache.fileCache.cacheKeyForKey(threeMileInstruction), cache.fileCache.cacheKeyForKey(sixMileInstruction))
```
Comment (akitchen) [inline MapboxNavigationTests/DataCacheTests.swift]: Great test case, this is the one that nails it.

File: MapboxNavigationTests/DataCacheTests.swift [hunk 4]
```diff
@@ -106,4 +106,27 @@ class DataCacheTests: XCTestCase {
...
+        XCTAssertEqual(cache.fileCache.cacheKeyForKey("-"), cache.fileCache.cacheKeyForKey("-"))
+    }
+    
+    func testCacheKeyPerformance() {
+        let instructionTurn = "Turn left"
+        let instructionContinue = "<speak><amazon:effect name=\"drc\"><prosody rate=\"1.08\">Continue on <say-as interpret-as=\"address\">I-80</say-as> East for 3 miles</prosody></amazon:effect></speak>"
+        measure {
+            for _ in 0...1000 {
+                _ = cache.fileCache.cacheKeyForKey(instructionTurn)
+                _ = cache.fileCache.cacheKeyForKey(instructionContinue)
```
Comment (1ec5) [inline MapboxNavigationTests/DataCacheTests.swift]: Can you post times from master and this branch for comparison?

File: MapboxNavigation/MD5.swift [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        output.append(value1)
+        output.append(value2)
+    }
+    
+    return String(output)
+}
+
+/*
+ * Convert a raw string to an array of little-endian words
+ * Characters >255 have their high-byte silently ignored.
```
Comment (1ec5) [inline MapboxNavigation/MD5.swift]: We do care about the high byte. Add test cases for non-ASCII characters, all the way up to emoji like 🛡 (U+1F6E1). Although emoji would be somewhat contrived for an OSM-based data set, there are plenty of multibyte characters in OSM, especially in Chinese and Japanese.

Comment (1ec5) [reply MapboxNavigation/MD5.swift]: Looks like we’re no longer using this code, but a test case would still be good for peace of mind.

File: MapboxNavigationTests/DataCacheTests.swift (hunk 4 above)
Comment (bsudekum) [reply MapboxNavigationTests/DataCacheTests.swift]: Yeah, it was 0.001 seconds for the same test.

**Pull request merged**